### PR TITLE
Simulator: add metadrive as optional simulator

### PR DESCRIPTION
**Pull request by jnewb1**
it currently works pretty well with cuda, but we should find a way to get the camera images quicker without using cuda

- updated to latest metadrive
- added metadrive sim
- refactored carla tests to also allow testing metadrive

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -132,7 +132,7 @@ inputs = "*"
 lru-dict = "*"
 markdown-it-py = "*"
 matplotlib = "*"
-metadrive-simulator = { git = "https://github.com/metadriverse/metadrive.git", rev ="51d4393f3b3574fd0e79ed04eae0081f8447ca72", markers = "platform_machine != 'aarch64'" } # no linux/aarch64 wheels for certain dependencies
+metadrive-simulator = { git = "https://github.com/metadriverse/metadrive.git", rev ="7d6f8ef707bfff67c6b88f3b4a98f8b1d58bf8e6", markers = "platform_machine != 'aarch64'" } # no linux/aarch64 wheels for certain dependencies
```
Comment (adeebshihadeh) [inline pyproject.toml:135]: can you put the update in its own PR?

File: tools/sim/bridge/metadrive.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    return img
+
+  def get_obs_as_rgb(self, buffer):
+    return (self.env.observations['default_agent'].img_obs.state * 255)[...,-1]
+
+  def apply_controls(self, steer_angle, throttle_out, brake_out):
+    steer_metadrive = steer_angle * 1 / (self.env.vehicle.MAX_STEERING * self.steer_ratio)
+    steer_metadrive = np.clip(steer_metadrive, -1, 1)
+
+    if (time.time() - self.reset_time) > 5:
```
Comment (adeebshihadeh) [inline tools/sim/bridge/metadrive.py:36]: use time.monotonic, time.time can jump

File: tools/sim/bridge/metadrive.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    self.steer_ratio = 15
+
+    self.vc = [0.0,0.0]
+
+    self.reset_time = 0
+
+  def get_cam_as_rgb(self, cam):
+    cam = self.env.engine.sensors[cam]
+    img = cam.perceive(self.env.vehicle, clip=False)
+    if type(img) != np.ndarray:
```
Comment (adeebshihadeh) [inline tools/sim/bridge/metadrive.py:25]: why would the type change?

File: tools/sim/bridge/metadrive.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  def tick(self):
+    obs, _, terminated, _, info = self.env.step(self.vc)
+
+    if terminated:
+      self.env.reset()
+      self.reset_time = time.time()
+
+  def close(self):
+    pass
```
Comment (adeebshihadeh) [inline tools/sim/bridge/metadrive.py:67]: this needed?

File: tools/sim/bridge/metadrive.py (hunk 3 above)
Comment (jnewb1) [reply tools/sim/bridge/metadrive.py:25]: if you are using cuda, you get a cupy array instead which is on the GPU, and needs to be moved

File: tools/sim/bridge/metadrive.py (hunk 4 above)
Comment (jnewb1) [reply tools/sim/bridge/metadrive.py:67]: implements an abstract class so it's needed

**Comment by adeebshihadeh** [comment]
Not sure if it's ready to try yet, but got this immediately
```
(openpilot-py3.11) batman:sim$ ./run_bridge.py --simulator metadrive
[INFO] MetaDrive version: 0.4.1.2
[INFO] Sensors: [lidar: Lidar(50,), side_detector: SideDetector(), lane_line_detector: LaneLineDetector(), rgb_road: RGBCameraRoad(1928, 1208), main_camera: MainCamera(1, 1)]
[INFO] Render Mode: offscreen
Process Process-1:
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/batman/openpilot/openpilot/tools/sim/bridge/common.py", line 63, in bridge_keep_alive
    self._run(q)
  File "/home/batman/openpilot/openpilot/tools/sim/bridge/common.py", line 84, in _run
    self.world = self.spawn_world()
                 ^^^^^^^^^^^^^^^^^^
  File "/home/batman/openpilot/openpilot/tools/sim/bridge/metadrive.py", line 121, in spawn_world
    env = MetaDriveEnv(
          ^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/metadrive_env.py", line 104, in __init__
    super(MetaDriveEnv, self).__init__(config)
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 255, in __init__
    self.agent_manager = self._get_agent_manager()
                         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 352, in _get_agent_manager
    return AgentManager(init_observations=self._get_observations(), init_action_space=self._get_action_space())
                                          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/metadrive_env.py", line 149, in _get_observations
    return {DEFAULT_AGENT: self.get_single_observation()}
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 603, in get_single_observation
    o = ImageStateObservation(self.config)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/obs/image_obs.py", line 25, in __init__
    self.img_obs = ImageObservation(config, config["vehicle_config"]["image_source"], config["rgb_clip"])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/obs/image_obs.py", line 50, in __init__
    assert _cuda_enable, "CuPy is not enabled"
AssertionError: CuPy is not enabled
```

**Comment by jnewb1** [comment]
> Not sure if it's ready to try yet, but got this immediately
> 
> ```
> (openpilot-py3.11) batman:sim$ ./run_bridge.py --simulator metadrive
> [INFO] MetaDrive version: 0.4.1.2
> [INFO] Sensors: [lidar: Lidar(50,), side_detector: SideDetector(), lane_line_detector: LaneLineDetector(), rgb_road: RGBCameraRoad(1928, 1208), main_camera: MainCamera(1, 1)]
> [INFO] Render Mode: offscreen
> Process Process-1:
> Traceback (most recent call last):
>   File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
>     self.run()
>   File "/home/batman/.pyenv/versions/3.11.4/lib/python3.11/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>   File "/home/batman/openpilot/openpilot/tools/sim/bridge/common.py", line 63, in bridge_keep_alive
>     self._run(q)
>   File "/home/batman/openpilot/openpilot/tools/sim/bridge/common.py", line 84, in _run
>     self.world = self.spawn_world()
>                  ^^^^^^^^^^^^^^^^^^
>   File "/home/batman/openpilot/openpilot/tools/sim/bridge/metadrive.py", line 121, in spawn_world
>     env = MetaDriveEnv(
>           ^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/metadrive_env.py", line 104, in __init__
>     super(MetaDriveEnv, self).__init__(config)
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 255, in __init__
>     self.agent_manager = self._get_agent_manager()
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 352, in _get_agent_manager
>     return AgentManager(init_observations=self._get_observations(), init_action_space=self._get_action_space())
>                                           ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/metadrive_env.py", line 149, in _get_observations
>     return {DEFAULT_AGENT: self.get_single_observation()}
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/envs/base_env.py", line 603, in get_single_observation
>     o = ImageStateObservation(self.config)
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/obs/image_obs.py", line 25, in __init__
>     self.img_obs = ImageObservation(config, config["vehicle_config"]["image_source"], config["rgb_clip"])
>                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/batman/.cache/pypoetry/virtualenvs/openpilot-TkpZrySJ-py3.11/lib/python3.11/site-packages/metadrive/obs/image_obs.py", line 50, in __init__
>     assert _cuda_enable, "CuPy is not enabled"
> AssertionError: CuPy is not enabled
> ```

currently needs a bunch of cuda dependencies, namely cupy and python-cuda to run locally

File: tools/sim/bridge/metadrive.py (hunk 2 above)
Comment (jnewb1) [reply tools/sim/bridge/metadrive.py]: done

**Comment by adeebshihadeh** [comment]
> I also plan to make it faster without cuda

For this particular use case, that would be great. MetaDrive is intended to be a solid, reliable baseline simulator experience for openpilot, so it would be nice if it didn't rely on too many HW-specific optimizations to just run at 20Hz on our workstations.

**Comment by jnewb1** [comment]
metadrive should launch now even if you don't have cuda. Performance isn't amazing but it at least engages and attempts to drive It probably makes sense to work on the performance in another PR

**Comment by adeebshihadeh** [comment]
Tried it out:
* took forever to load initially. seems MetaDrive loads stuff on the first run? can we show progress or at least print that this is expected?
* manager is running way too fast, should be 2Hz
* overall, sim quality and frame rate were good on my machine!

**Comment by jnewb1** [comment]
- added a print statement saying it might take awhile to load initially, I believe this is due to model compression, but that helps low-memory machines. We can probably default to no compression, and add a flag to enable it for low memory machines
- was sending pandastates way too fast which caused the manager to run at the wrong frequency, fixed that
